File: src/contest/spoj/DynamicLca.java
package contest.spoj;

import java.util.*;
import java.io.*;

public class DynamicLca {
	private static Node[] nodes;

	private static class Node {
		Node path_parent;
		// tree pointers
		Node parent, left, right;
		Integer id;
		Node (Integer id) {
			this.id = id;
		}
	}
	// precondition: n must be a root node, and n and m must be in different trees
	public static void link (Node n, Node m) {
		access(n);
		access(m);

		n.left = m;
		m.parent = n;
		m.path_parent = null;
	}
	// precondition: n is not a root node
	public static void cut (Node n) {
		access(n);
		if (n.left != null) {
			n.left.parent = null;
			n.left.path_parent = null;
			n.left = null;
		}
	}

	public static Node getRoot (Node n) {
		access(n);
		while (n.left != null)
			n = n.left;
		splay(n);
		return n;
	}

	public static void access (Node n) {
		splay(n);
		if (n.right != null) {
			n.right.path_parent = n;
			n.right.parent = null;
			n.right = null;
		}
		Node m = n;
		while (m.path_parent != null) {
			Node nextTree = m.path_parent;
			splay(nextTree);
			if (nextTree.right != null) {
				nextTree.right.path_parent = nextTree;
				nextTree.right.parent = null;
			}
			nextTree.right = m;
			m.parent = nextTree;
			m.path_parent = null;

			m = nextTree;
		}
		splay(n);
	}
	private static void splay (Node n) {
		while (n.parent != null) {
			Node p = n.parent;
			Node pp = n.parent.parent;
			if (pp == null) {
				rotate(n);
			} else if ((pp.left == p && p.left == n) || (pp.right == p && p.right == n)) {
				rotate(p);
				rotate(n);
			} else {
				rotate(n);
				rotate(n);
			}
		}
	}
	private static void rotate (Node n) {
		Node p = n.parent;
		Node pp = n.parent.parent;
		if (p.left == n) {
			p.left = n.right;
			n.right = p;
			p.parent = n;
			if (p.left != null)	
				p.left.parent = p;
		} else if (p.right == n) {
			p.right = n.left;
			n.left = p;
			p.parent = n;
			if (p.right != null)
				p.right.parent = p;
		}
		n.parent = pp;
		if (pp != null) {
			if (pp.left == p)
				pp.left = n;
			else
				pp.right = n;
		}
		n.path_parent = p.path_parent;
		p.path_parent = null;
	}

	public static Node lca (Node n, Node m) {
		if (getRoot(n) != getRoot(m))
			return null;
		access(m);
		splay(n);
		if (n.right != null) {
			n.right.path_parent = n;
			n.right.parent = null;
			n.right = null;
		}
		Node res = n;
		Node k = n;
		while (k.path_parent != null) {
			Node nextTree = k.path_parent;
			splay(nextTree);
			if (nextTree.path_parent == null)
				res = nextTree;
			if (nextTree.right != null) {
				nextTree.right.path_parent = nextTree;
				nextTree.right.parent = null;
			}
			nextTree.right = k;
			k.parent = nextTree;
			k.path_parent = null;

			k = nextTree;
		}
		splay(n);
		return res;
	}

	static BufferedReader br;
	static PrintWriter out;
	static StringTokenizer st;
	public static void main (String[] args) throws IOException {
		br = new BufferedReader(new InputStreamReader(System.in));
		out = new PrintWriter(new OutputStreamWriter(System.out));
		
		int n = readInt();
		int m = readInt();
		nodes = new Node[n];
		for (int i = 0; i < n; i++)
			nodes[i] = new Node(i+1);
		
		for (int i = 0; i < m; i++) {
			String command = next();
			if (command.equals("link")) {
				int j = readInt()-1;
				int k = readInt()-1;
				link(nodes[j], nodes[k]);
			} else if (command.equals("cut")) {
				int j = readInt()-1;
				cut(nodes[j]);
			} else {
				int j = readInt()-1;
				int k = readInt()-1;
				Node lca = lca(nodes[j], nodes[k]);
				out.println(lca.id);
			}
		}

		out.close();
	}
	static String next () throws IOException {
		while (st == null || !st.hasMoreTokens())
			st = new StringTokenizer(br.readLine().trim());
		return st.nextToken();
	}

	static long readLong () throws IOException {
		return Long.parseLong(next());
	}

	static int readInt () throws IOException {
		return Integer.parseInt(next());
	}

	static double readDouble () throws IOException {
		return Double.parseDouble(next());
	}

	static char readCharacter () throws IOException {
		return next().charAt(0);
	}

	static String readLine () throws IOException {
		return br.readLine().trim();
	}
}

File: src/contest/spoj/DynamicLca.cc
#include <bits/stdc++.h>

using namespace std;

struct LinkCut {
  struct Node {
    int index, sz;
    Node *child[2], *par, *pathPar;
    Node (int index, int sz): index(index), sz(sz) {
      child[0] = this;
      child[1] = this;
      par = this;
      pathPar = this;
    }
  };

  static Node *null;
  vector<Node*> nodes;

  LinkCut (int N): nodes(N) {
    for (int i = 0; i < N; i++) {
      nodes[i] = new Node(i + 1, 1);
      nodes[i]->child[0] = nodes[i]->child[1] = nodes[i]->par = nodes[i]->pathPar = null;
    }
  }

  static void update (Node *u) {
    if (u == null)
      return;
    u->sz = 1 + u->child[0]->sz + u->child[1]->sz;
  }

  static int getDir (Node *u, Node *p) {
    return p->child[0] == u ? 0 : 1;
  }

  static void connect (Node* u, Node* v, int dir) {
    u->child[dir] = v;
    v->par = u;
  }

  static Node* rotate (Node* u, int dir) {
    Node *c = u->child[dir ^ 1], *p = u->par, *pp = p->par;
    connect(p, c, dir);
    connect(u, p, dir ^ 1);
    connect(pp, u, getDir(p, pp));

    u->pathPar = p->pathPar;
    p->pathPar = null;
    update(p);
    update(u);
    update(pp);
    return u;
  }

  static void splay (Node *u) {
    while (u->par != null) {
      Node *p = u->par, *pp = p->par;
      int dp = getDir(u, p), dpp = getDir(p, pp);
      if (pp == null) rotate(u, dp);
      else if (dp == dpp) rotate(p, dpp), rotate(u, dp);
      else rotate(u, dp), rotate(u, dpp);
    }
  }

  static Node* access (Node* u) {
    Node *prev = null;
    for (Node *v = u; v != null; v = v->pathPar) {
      splay(v);
      if (v->child[1] != null) {
        v->child[1]->pathPar = v;
        v->child[1]->par = null;
        v->child[1] = null;
      }
      v->child[1] = prev;
      update(v);
      if (prev != null) {
        prev->par = v;
        prev->pathPar = null;
      }
      prev = v;
    }
    splay(u);
    return prev;
  }

  // precondition: n must be a root node, and n and m must be in different trees
  static void link (Node *n, Node *m) {
    access(n);
    access(m);
    n->child[0] = m;
    m->par = n;
    update(n);
  }

  // precondition: n must not be a root node
  static void cut (Node *n) {
    access(n);
    if (n->child[0] != null) {
      n->child[0]->par = null;
      n->child[0] = null;
    }
    update(n);
  }

  static Node* getRoot (Node *n) {
    access(n);
    while (n->child[0] != null)
      n = n->child[0];
    access(n);
    return n;
  }

  static int getHeight (Node *n) {
    access(n);
    return n->child[0]->sz + 1;
  }

  static int lca (Node *u, Node *v) {
    access(u);
    return access(v)->index;
  }
};

LinkCut::Node* LinkCut::null = new Node(0, 0);

int main () {
  int N, M;
  cin >> N >> M;

  LinkCut t(N);
  for (int i = 0; i < M; i++) {
    string in;
    cin >> in;
    int u, v;
    if (in[1] == 'c') {
      cin >> u >> v;
      cout << t.lca(t.nodes[u - 1], t.nodes[v - 1]) << "\n";
    } else if (in[1] == 'i') {
      cin >> u >> v;
      t.link(t.nodes[u - 1], t.nodes[v - 1]);
    } else {
      cin >> u;
      t.cut(t.nodes[u - 1]);
    }
  }
  return 0;
}
